lib/Conversion/LLHDToLLVM/LLHDToLLVM.cpp

```diff
@@ -29,13 +29,22 @@ using namespace circt::llhd;
 
 // Keep a counter to infer a signal's index in his entity's signal table.
 static size_t signalCounter = 0;
-// Keep a counter to infer the resume index after a wait instruction in a
-// process.
-static int resumeCounter = 0;
+
+// Keep a counter to infer a reg's index in his entity.
+static size_t regCounter = 0;
```

Comment:
Mutable global state is generally problematic in MLIR (e.g. due to multithreading), I'd recommend moving this to per-instance state in the pass.  Similarly with signalCounter.

Reply:
Agreed, I never really liked this hacky approach but was never sure of what is a good way to make per-instance state available to the patterns. Can they just be stored as members of the pass and references fed to the patterns constructors?

Reply:
I made them local to `runOnOperation()` and make them available to the patterns when inserting them in the pattern list. I'm pretty sure there's a more elegant way of doing it but it should at least be more solid.

Reply:
I think that's a great start. We can always refactor in the future if we come up with a better solution; but at least there are no more data races :+1:

Reply:
Yeah I think it could be a good idea to have some kind of analysis that can be shared between the lowering pass and the engine at some point, which guarantees the indexing to be done in the same order for both instead of relying on counting (and would also replace the current initial state construction walk)!

Reply:
A later refactoring :wink:!